In [25]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
from google.colab import files
import warnings
warnings.filterwarnings('ignore')

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [26]:
#importing CSV uploaded to drive
id = "1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v"
print("https://drive.google.com/uc?export=download&id=" + id)
!wget "https://drive.google.com/uc?export=download&id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v" -O Online_Retail.csv

https://drive.google.com/uc?export=download&id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v
--2026-01-15 07:17:31--  https://drive.google.com/uc?export=download&id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v
Resolving drive.google.com (drive.google.com)... 108.177.125.113, 108.177.125.102, 108.177.125.138, ...
Connecting to drive.google.com (drive.google.com)|108.177.125.113|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://drive.usercontent.google.com/download?id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v&export=download [following]
--2026-01-15 07:17:31--  https://drive.usercontent.google.com/download?id=1M5IPR96R9efi5cb2n8LxpXIXyx_F0x3v&export=download
Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 74.125.23.132, 2404:6800:4008:c1b::84
Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|74.125.23.132|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 45005763 (43M) [application/octet-stream]
Saving to:

In [27]:
df=pd.read_csv('/content/Online_Retail.csv')
df.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [28]:
print(df.info())
print(df.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB
None
(541909, 8)


In [29]:
#Given quantity and UnitPrice can't be negative, removing negative values

df=df[(df['Quantity']>=0) & (df['UnitPrice']>=0)]
print(df.shape)
df.head()

(531283, 8)


InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [30]:
#Count  of customers, count of unique invoices
print(df['CustomerID'].nunique())
print(df['InvoiceNo'].nunique())

4339
20726


In [31]:
df['Country'].value_counts()

Country
United Kingdom          486284
Germany                   9042
France                    8408
EIRE                      7894
Spain                     2485
Netherlands               2363
Belgium                   2031
Switzerland               1967
Portugal                  1501
Australia                 1185
Norway                    1072
Italy                      758
Channel Islands            748
Finland                    685
Cyprus                     614
Sweden                     451
Unspecified                446
Austria                    398
Denmark                    380
Poland                     330
Japan                      321
Israel                     295
Hong Kong                  284
Singapore                  222
Iceland                    182
USA                        179
Canada                     151
Greece                     145
Malta                      112
United Arab Emirates        68
European Community          60
RSA                         58
Lebanon                     45
Lithuania                   35
Brazil                      32
Czech Republic              25
Bahrain                     18
Saudi Arabia                 9
Name: count, dtype: int64

In [32]:
# Given that majority of the invoices are from UK and different countries
# have different item sets, filtering only UK data

data=df[df['Country']=='United Kingdom']
print(data.shape)
print(data.head(3))

(486284, 8)
  InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom

[3 rows x 8 columns]


In [33]:
data['Description'].nunique()

4058

In [34]:
# Grouping on Invoice + description level and sum of quantity
#  Pivoting the data
#  Creating binary values

data=data.groupby(['InvoiceNo','Description'])['Quantity'].sum()
# data.head(3)
data=data.unstack().reset_index().fillna(0)
data.head()

Description InvoiceNo  ...  wrongly sold (22719) barcode
0              536365  ...                           0.0
1              536366  ...                           0.0
2              536367  ...                           0.0
3              536368  ...                           0.0
4              536369  ...                           0.0

[5 rows x 4059 columns]

In [35]:
#Create binary values
data=data.set_index('InvoiceNo')
data.head()

Description   4 PURPLE FLOCK DINNER CANDLES  ...  wrongly sold (22719) barcode
InvoiceNo                                    ...                              
536365                                  0.0  ...                           0.0
536366                                  0.0  ...                           0.0
536367                                  0.0  ...                           0.0
536368                                  0.0  ...                           0.0
536369                                  0.0  ...                           0.0

[5 rows x 4058 columns]

In [37]:
def encode_quantity(x):
  if x<=0:
    return 0
  else:
    return 1

data=data.applymap(encode_quantity)
data.tail()

Description   4 PURPLE FLOCK DINNER CANDLES  ...  wrongly sold (22719) barcode
InvoiceNo                                    ...                              
581583                                    0  ...                             0
581584                                    0  ...                             0
581585                                    0  ...                             0
581586                                    0  ...                             0
A563185                                   0  ...                             0

[5 rows x 4058 columns]

In [40]:
# Filter invoices which contain atleast two different items

data=data[(data>0).sum(axis=1)>=2]
data.shape

(16539, 4058)

In [41]:
frequent_itemsets_plus=apriori(data,min_support=0.03,
                               use_colnames=True).sort_values('support',ascending=False).reset_index(drop=True)
frequent_itemsets_plus['length']=frequent_itemsets_plus['itemsets'].apply(lambda x:len(x))
frequent_itemsets_plus

support                                           itemsets  length
0    0.129875               (WHITE HANGING HEART T-LIGHT HOLDER)       1
1    0.116331                          (JUMBO BAG RED RETROSPOT)       1
2    0.100671                         (REGENCY CAKESTAND 3 TIER)       1
3    0.095471                                    (PARTY BUNTING)       1
4    0.084165                          (LUNCH BAG RED RETROSPOT)       1
..        ...                                                ...     ...
172  0.030473                   (CHRISTMAS CRAFT LITTLE FRIENDS)       1
173  0.030473                          (CREAM HEART CARD HOLDER)       1
174  0.030413     (LUNCH BAG  BLACK SKULL., LUNCH BAG CARS BLUE)       2
175  0.030292  (LUNCH BAG SPACEBOY DESIGN , LUNCH BAG RED RET...       2
176  0.030171  (LUNCH BAG  BLACK SKULL., LUNCH BAG SUKI DESIGN )       2

[177 rows x 3 columns]

In [42]:
# Association Rules
asc_rules=association_rules(frequent_itemsets_plus,metric='lift',min_threshold=1)
asc_rules=asc_rules.sort_values('lift',ascending=False).reset_index(drop=True)
asc_rules.head()

antecedents  ... kulczynski
0    (GREEN REGENCY TEACUP AND SAUCER)  ...   0.721619
1     (PINK REGENCY TEACUP AND SAUCER)  ...   0.721619
2     (PINK REGENCY TEACUP AND SAUCER)  ...   0.676721
3   (ROSES REGENCY TEACUP AND SAUCER )  ...   0.676721
4  (GARDENERS KNEELING PAD KEEP CALM )  ...   0.662709

[5 rows x 14 columns]

## Summary

This notebook performs an association rule mining analysis on an Online Retail dataset to discover frequently co-occurring items and derive association rules. The analysis focuses specifically on sales data from the United Kingdom.

### 1. Initial Setup and Data Loading

The notebook begins by importing necessary libraries such as `pandas`, `numpy`, `matplotlib`, `seaborn`, `mlxtend.frequent_patterns` (for Apriori and association rules), and `warnings`. The dataset, `Online_Retail.csv`, is downloaded from Google Drive using a `wget` command and then loaded into a pandas DataFrame.

### 2. Data Inspection and Preprocessing

- **Initial Inspection**: The `df.info()` and `df.shape` methods are used to get an overview of the dataset's structure, data types, and missing values. The dataset contains 541,909 entries and 8 columns.
- **Data Cleaning**: Rows with negative `Quantity` or `UnitPrice` are removed, as these are typically invalid in sales data. This reduces the dataset to 531,283 entries.
- **Customer and Invoice Counts**: The number of unique customers (4339) and invoices (20726) are identified.
- **Country Filtering**: The `Country` column is analyzed, revealing that the United Kingdom accounts for the vast majority of transactions. To simplify the analysis and focus on a specific market, the dataset is filtered to include only transactions from the 'United Kingdom', resulting in 486,284 entries.

### 3. Data Transformation for Association Rule Mining

- **Aggregation**: The filtered UK data is grouped by `InvoiceNo` and `Description`, and the `Quantity` is summed up for each unique invoice-item combination.
- **Pivoting (One-Hot Encoding Preparation)**: The grouped data is then unstacked, creating a matrix where rows are `InvoiceNo` and columns are `Description` (items), with `Quantity` as values. Missing values introduced by unstacking are filled with 0.
- **Binary Encoding**: A custom function `encode_quantity` is applied to transform the quantities into binary values (1 if an item was purchased in an invoice, 0 otherwise). The `InvoiceNo` column is set as the index.
- **Filtering Invoices**: Invoices containing less than two distinct items are removed, as they are not relevant for finding associations between items.

### 4. Apriori Algorithm and Association Rules

- **Frequent Itemsets**: The `apriori` algorithm from `mlxtend.frequent_patterns` is applied to the binary-encoded data to find frequent itemsets. A `min_support` of 0.03 (3%) is used, and itemsets are sorted by support in descending order. A 'length' column is added to indicate the number of items in each set. This step identifies combinations of items that frequently appear together in transactions.
- **Association Rules Generation**: Association rules are then generated from the frequent itemsets using the `association_rules` function. The `metric` is set to 'lift', and `min_threshold` to 1, focusing on rules where the purchase of the antecedent positively influences the purchase of the consequent. The rules are sorted by 'lift' in descending order. This step reveals actionable insights, such as which items are likely to be bought together.

### 5. Conclusion

The notebook successfully processed a large retail dataset, transformed it into a suitable format for association rule mining, and applied the Apriori algorithm to discover frequent itemsets and significant association rules within the United Kingdom's sales data. The resulting `asc_rules` DataFrame provides insights into product relationships that can be used for recommendations, store layout optimization, or targeted marketing strategies.